keps/sig-node/4381-dra-numeric-parameters/README.md

```diff
@@ -0,0 +1,1054 @@
+  to ensure that configuration parameters are available when the DRA driver
+  needs them.
+
+The types that the scheduler needs to watch may change at runtime as classes
+get added or removed. To avoid repeated GET operations for those CRDs, dynamic
+informers will have to be used that get started and stopped as needed.
+
+## Design Details
+
+### ResourceClass extension
+
+A new, optional field in ResourceClass enables semantic parameters for claims
+using this class:
+
+```
+type ResourceClass struct {
+	...
+
+	// If (and only if) allocation of claims using this class is handled
+	// via numeric parameters, then ParameterTypes describes those
+	// parameters. A driver might support multiple different CRDs for
+	// its claims.
+	ParameterTypes []SemanticParameterType
+}
+
+// SemanticParameterType describes one resource that may be used for
+// claim parameters of a class which uses semantic parameters.
+type SemanticParameterType struct {
+	// APIGroup is the group for the resource being referenced. It is
+	// empty for the core API. This matches the group in the APIVersion
+	// that is used when creating the resources.
+	// +optional
+	APIGroup string
+	// Kind is the type of resource being referenced. This is the same
+	// value as in the parameter object's metadata.
+	Kind string
+
+	// Shareable is copied into the AllocationResult when a claim
+	// gets allocated.
+	Shareable bool
+}
+```
+
+### NodeResources
+
+For each node, one or more NodeResources objects gets created:
+
+```
+// NodeResources provides information about available
+// resources on individual nodes.
+type NodeResources struct {
+	metav1.TypeMeta
+	// Standard object metadata
+	// +optional
+	metav1.ObjectMeta
+
+	// NodeName identifies the node where the capacity is available.
+	// A field selector can be used to list only NodeResources
+	// objects with a certain node name.
+	NodeName string
+
+	// DriverName identifies the DRA driver providing the capacity information.
+	// A field selector can be used to list only NodeResources
+	// objects with a certain driver name.
+	DriverName string
+
+	// Instances describes discrete resources managed by the driver on
+	// the node. It is possible to create more than one NodeResources
+	// object for the same node and driver if this array becomes to large
+	// for a single object.
+	Instances []NodeResourceInstance
+}
+
+// NodeResourceInstance describes one discrete resource instance.
+type NodeResourceInstance struct {
+	// ID is chosen by the driver to distinguish between different resource
+	// instances. It only needs to be unique for the driver and node.  In
+	// other words, the tuple of NodeName, DriverName, InstanceID needs to
+	// be unique in the cluster.
+	ID string
+
+    // Attributes describe the instance.
+	Attributes []InstanceAttributes
+}
+
+// InstanceAttributes must have one and only one field set.
+type InstanceAttributes struct {
+	<one pointer per semantic model>
```

Comment:
I don't understand why semantic models come into play at this level. An instance should have a set of attributes that can be selected on, period. I would actually expect the whole notion of semantic models to be **_completely decoupled_** from the `NodeResources` object. 

Where the `NodeResources` object declares the full set of possible resources that *could* be available (at all), and something else stitches together the way in which the resources declared here might be allocated out (i.e. according to different "semantic models")..

Reply:
Modeling driver resources as "finite set of instances" with one counter for the set and some non-quantitative attributes for each instance *is* already a semantic model. It needs to be expressed in the `NodeResources` type. There might be other, more complex resources which need different resource model and thus representation. 

Therefore `NodeResources` depends on semantic models.